tensorflow/compiler/xla/service/gpu/ir_emitter.cc

```diff
@@ -371,12 +364,21 @@ Status IrEmitter::EmitAtomicOperationUsingCAS(const HloComputation& computation,
     llvm::Type* address_int_type =
         module_->getDataLayout().getIntPtrType(output_address_type);
     atomic_memory_address = PtrToInt(output_address, address_int_type);
-    llvm::Value* mask = llvm::ConstantInt::get(address_int_type, -4);
+    llvm::Value* mask = llvm::ConstantInt::get(address_int_type, 3);
+    llvm::Value* offset = And(atomic_memory_address, mask);
+    mask = llvm::ConstantInt::get(address_int_type, -4);
     atomic_memory_address = And(atomic_memory_address, mask);
     atomic_memory_address =
         IntToPtr(atomic_memory_address, atomic_address_type);
-    binop_output_address = PointerBitCastOrAddrSpaceCast(cas_new_output_address,
-                                                         element_address_type);
+    binop_output_address =
+        Add(PtrToInt(cas_new_output_address, address_int_type), offset);
+    llvm::Type* cas_new_output_pointer_type = cas_new_output_address->getType();
+    CHECK_EQ(true, cas_new_output_pointer_type->isPointerTy());
+    unsigned correct_address_space =
+        cas_new_output_pointer_type->getPointerAddressSpace();
```

Comment:
`cas_new_output_pointer_type` is not necessary as LLVM alloca always returns a pointer.

so line 375 - line 378 could be absorbed into the following line as:

```
	    binop_output_address =
        IntToPtr(binop_output_address,
                 llvm::PointerType::get(element_type, cas_new_output_address->getType()->getPointerAddressSpace()));
```

Reply:
to the very least line 375 - 376 shall be eliminated.